network/requestLogger.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (C) 2019 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"bufio"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/logging/telemetryspec"
+)
+
+// RequestLogger is a middleware helps logging all the incoming http requests.
+// The intended use is to place it at the bottom of the http processing. It will capture the status codes
+// set by the upsteam handlers and write the request info/response to the logger.
+type RequestLogger struct {
+	downsteamHandler    http.Handler
+	trackingWritersPool sync.Pool
+	log                 logging.Logger
+}
+
+// create a RequestLogger object.
+func makeRequestLogger(downsteamHandler http.Handler, log logging.Logger) *RequestLogger {
+	rl := &RequestLogger{
+		downsteamHandler: downsteamHandler,
+		log:              log,
+	}
+	rl.trackingWritersPool = sync.Pool{
+		New: func() interface{} {
+			return &trackingResponseWriter{}
+		},
+	}
+	return rl
+}
+
+// this is the http entry point for the request logger.
+func (rl *RequestLogger) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	trackingWriter := rl.trackingWritersPool.Get().(*trackingResponseWriter)
+	trackingWriter.Reset(writer)
+	defer func() {
```

Comment:
Is there a reason we need to defer this rather than just execute it after the call to ServeHTTP()?  In practice, we don't recover from panics.

Reply:
Interesting point. I'm not sure it's correct. if you'll look into server.go in the go codebase, you'll see the following:
```
// Serve a new connection.
func (c *conn) serve(ctx context.Context) {
	c.remoteAddr = c.rwc.RemoteAddr().String()
	ctx = context.WithValue(ctx, LocalAddrContextKey, c.rwc.LocalAddr())
	defer func() {
		if err := recover(); err != nil && err != ErrAbortHandler {
			const size = 64 << 10
			buf := make([]byte, size)
			buf = buf[:runtime.Stack(buf, false)]
			c.server.logf("http: panic serving %v: %v\n%s", c.remoteAddr, err, buf)
		}
		if !c.hijacked() {
			c.close()
			c.setState(c.rwc, StateClosed)
		}
	}()
...
```

it looks to me as if it would print the error and close the connection.

Reply:
What I meant is that _we_ (above ServeHTTP on the stack) do not recover, so the defer() will only be useful if downsteamHandler.ServeHTTP() panics; but we'll end up exiting anyway because we don't recover.  Since the panic() is simply logged, it means our defer() isn't required to ensure that block of code runs.  Not a big concern, just wondering.